### PR TITLE
add draft status systemfield

### PR DESCRIPTION
**Pull request by zzacharo**
:heart: Thank you for your contribution!

### Description

Please describe briefly your pull request.

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [ ] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [ ] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [ ] I've added relevant test cases.
- [ ] I've added relevant documentation.
- [ ] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code).
- [ ] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code).
- [ ] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code).
- [ ] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [ ] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [ ] I've NOT included third-party code (copy/pasted source code or new dependencies).

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_rdm_records/records/systemfields/draft_status.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        if record is None:
+            return self  # returns the field itself.
+        parent = getattr(record, 'parent')
+        review = getattr(parent, 'review')
+        is_published = getattr(record, 'is_published')
+
+        if is_published:
+            return self.available_draft_status["published"]
+        if review is None:
+                return self.available_draft_status["draft"]
```
Comment (ntarocco) [inline invenio_rdm_records/records/systemfields/draft_status.py:70]:
```suggestion
        return self.available_draft_status["draft"]
```

File: invenio_rdm_records/records/systemfields/draft_status.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    def __get__(self, record, owner=None):
+        """Get the persistent identifier."""
+        if record is None:
+            return self  # returns the field itself.
+        parent = getattr(record, 'parent')
+        review = getattr(parent, 'review')
+        is_published = getattr(record, 'is_published')
+
+        if is_published:
+            return self.available_draft_status["published"]
```
Comment (ntarocco) [inline invenio_rdm_records/records/systemfields/draft_status.py:68]: is `available_draft_status` defined?

File: invenio_rdm_records/requests/community_submission.py [hunk 3]
```diff
@@ -69,13 +69,16 @@ class DeclineAction(actions.DeclineAction):
...
-        # This means that the receiver (curator) won't have access anymore
+        # Keeps the record and the review connected so the user can see the
+        # outcome of the request
+        # The receiver (curator) won't have access anymore to the draft
         # The creator (uploader) should still have access to the record/draft
         draft = self.request.topic.resolve()
-        draft.parent.review = None
         super().execute(identity, uow)
 
+        draft.parent.review = self.request
```
Comment (lnielsen) [inline invenio_rdm_records/requests/community_submission.py:79]: Minor: It ought to already be assigned, so it seems strange this line is neeeded. You'll still need the line below (``RecordCommitOp``), but can you move it before the ``super().execute()``

File: invenio_rdm_records/requests/community_submission.py [hunk 4]
```diff
@@ -95,15 +98,16 @@ class ExpireAction(actions.CancelAction):
...
-        # Remove draft from request
         # Same reasoning as in 'decline'
         draft = self.request.topic.resolve()
-        draft.parent.review = None
 
         # TODO: What more to do? simply close the request? Similarly to
         # decline, how does a user resubmits the request to the same community.
         super().execute(identity, uow)
 
+        draft.parent.review = self.request
```
Comment (lnielsen) [inline invenio_rdm_records/requests/community_submission.py:108]: Ditto here

File: invenio_rdm_records/services/review/service.py [hunk 5]
```diff
@@ -123,12 +123,15 @@ def delete(self, identity, id_, revision_id=None, uow=None):
...
-        # Delete request
-        current_requests_service.delete(
-            identity,
-            draft.parent.review.id,
-            uow=uow
-        )
+        # Delete request if request not in ['expired', 'declined']
+        # we keep the request in this case so the user can see the outcome of
+        # the request
+        if draft.parent.review.status not in ['expired', 'declined']:
```
Comment (lnielsen) [inline invenio_rdm_records/services/review/service.py:129]: Minor: I think it's safer to use the ``is_closed/is_open`` properties here. You can only delete a request in ``created`` state (which is neither closed nor open):

```suggestion
        if not (draft.parent.review.is_closed or draft.parent.review.is_open):
```

File: invenio_rdm_records/services/review/service.py [hunk 6]
```diff
@@ -146,8 +149,11 @@ def submit(self, identity, id_, data=None, revision_id=None, uow=None):
...
             raise ReviewNotFoundError()
 
+
         # All other preconditions can be checked by the action itself which can
         # raise appropriate exceptions.
-        request = current_requests_service.execute_action(
+        request_item = current_requests_service.execute_action(
             identity, draft.parent.review.id, 'submit', data=data, uow=uow)
-        return request
+        draft.parent.review = request_item._request
```
Comment (lnielsen) [inline invenio_rdm_records/services/review/service.py:157]: Ditto to why this assignment is needed?

File: invenio_rdm_records/records/api.py [hunk 7]
```diff
@@ -228,6 +229,8 @@ class RDMDraft(CommonFieldsMixin, Draft):
 
     has_draft = HasDraftCheckField()
 
+    status = DraftStatus()
```
Comment (lnielsen) [inline invenio_rdm_records/records/api.py:232]: I think we need a status field on the record as well for the index. It's probably a constant field simply saying ``published``.

File: invenio_rdm_records/records/systemfields/draft_status.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+from invenio_records.systemfields import SystemField
+
+from invenio_rdm_records.services.errors import ReviewStateError
+
+
+class DraftStatus(SystemField):
+    """Draft status field which checks the `parent.review` state."""
+
+    available_review_to_draft_status = dict(
+        draft=_("Draft"),
```
Comment (lnielsen) [inline invenio_rdm_records/records/systemfields/draft_status.py:34]: Question: This one is a bit tricky. The "list uploads" page needs to display a facet over this status. The facet I think should display:

- Draft: draft + draft with review states 
- In review
- Review expired 
- Review declined
- Published

This assumes end-users won't understand the difference between ``Draft`` and ``Draft with review``. If we think end-users will understand, and/or that users want to find all drafts being submitted for review then we can leave as is.

File: invenio_rdm_records/records/systemfields/draft_status.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+"""
+
+from flask_babelex import gettext as _
+from invenio_records.systemfields import SystemField
+
+from invenio_rdm_records.services.errors import ReviewStateError
+
+
+class DraftStatus(SystemField):
```
Comment (lnielsen) [inline invenio_rdm_records/records/systemfields/draft_status.py:30]: I think we need to factor in both the record and draft type as well (``new draft``, ``edit draft``,  ``next draft``) or we need some other field specifying the draft type.

File: invenio_rdm_records/services/review/service.py (hunk 6 above)
Comment (zzacharo) [reply invenio_rdm_records/services/review/service.py:156]: Add comment as above for this assignment

**Pull request merged**